test/e2e/framework/expect.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2021 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/tools/cache"
+)
+
+// Expectation closes over a statement of intent, allowing the caller
+// to accumulate errors and determine when the expectation should cease
+// to be evaluated.
+type Expectation func() (done bool, err error)
+
+// Expecter allows callers to register expectations
+type Expecter interface {
+	// ExpectBefore will result in the Expectation being evaluated whenever
+	// state changes, up until the desired timeout is reached.
+	ExpectBefore(Expectation, time.Duration)
+}
+
+func NewExpecter(informer cache.SharedIndexInformer) *expectationController {
+	controller := expectationController{
+		informer:     informer,
+		expectations: map[uuid.UUID]func(){},
+		lock:         &sync.RWMutex{},
+	}
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(_ interface{}) {
+			controller.triggerExpectations()
+		},
+		UpdateFunc: func(_, _ interface{}) {
+			controller.triggerExpectations()
+		},
+		DeleteFunc: func(_ interface{}) {
+			controller.triggerExpectations()
+		},
+	})
+
+	return &controller
+}
+
+type expectationController struct {
+	informer     cache.SharedIndexInformer
+	expectations map[uuid.UUID]func()
+	lock         *sync.RWMutex
+}
+
+func (c *expectationController) triggerExpectations() {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	for _, expectation := range c.expectations {
+		expectation()
+	}
+}
+
+func (c *expectationController) ExpectBefore(expectation Expectation, duration time.Duration) error {
```

Comment:
```suggestion
func (c *expectationController) ExpectBefore(ctx context.Context, expectation Expectation, duration time.Duration) error {
```

When you're calling this method from test cases, they already have a context. I think it would make sense to take in that context so this code can react to the passed in context being done, if needed. WDYT?